stdlib/atexit.c

```diff
@@ -87,28 +89,32 @@ static int _atexit_register(int isarg, void (*fn)(void), void *arg)
 /* Call all registered destructors */
 void _atexit_call(void)
 {
-	struct atexit_node *last, *node = atexit_common.head;
-	int i;
-
-	while (node != NULL) {
-		for (i = ATEXIT_MAX - 1; i >= 0; i--) {
-			if (node->destructors[i] != NULL) {
-				if (((node->fntype >> i) & 1) == 1) {
-					((void (*)(void *))node->destructors[i])(node->args[i]);
-				}
-				else {
-					node->destructors[i]();
-				}
-			}
+	mutexLock(atexit_common.lock);
+	while ((atexit_common.head != NULL) && (atexit_common.idx != 0)) {
+		atexit_common.idx--;
+		destructor_t destructor = atexit_common.head->destructors[atexit_common.idx];
+		if (((atexit_common.head->fntype >> atexit_common.idx) & 1) == 1) {
+			void *arg = atexit_common.head->args[atexit_common.idx];
+			mutexUnlock(atexit_common.lock);
+			((void (*)(void *))destructor)(arg);
 		}
-
-		last = node;
-		node = node->prev;
-		if (node != NULL) {
-			/* Free only if it's not the first node (statically allocated) */
-			free(last);
+		else {
+			mutexUnlock(atexit_common.lock);
+			destructor();
+		}
+		mutexLock(atexit_common.lock);
+
+		if (atexit_common.idx == 0) {
+			struct atexit_node *last = atexit_common.head;
+			atexit_common.head = atexit_common.head->prev;
+			if (atexit_common.head != NULL) {
+				/* Free only if it's not the first node (statically allocated) */
```

Comment:
This whole thing with first node being allocated statically is kinda a bad thing imho. We lose reference to it here (I guess no big deal, as we exit anyway). I'd consider lazy allocation (on first register ) instead. Just a suggestion to consider

Reply:
Yeah, good idea. Just take note that the first `atexit()` calls can be done before main() - I think as a part of init array (C++ global objects). 

Side note: this is a good example of PR after which some C++ testing would be welcome (ensuring we won't break  C++ destructors for object of any storage class).

Reply:
Ok, so this might be a reason for this solution being used here. This code was added as a part of C++ support, this further support this argument. So my idea would break C++